### PR TITLE
8279292: riscv: Intrinsify multiplyToLen and squareToLen

### PR DESCRIPTION
**Pull request by feilongjiang**
BigInteger intrinsic: MultiplyToLen and SquareToLen intrinsic are missed in current vm. They should be implemented.
The JMH tests show that the MultiplyToLen intrinsic improve the performance by up to 2x ~ 3x and the SquareToLen intrinsic improve the performance by up to 1.8x ~ 2x when the length of BigInteger changed from 1 to 5000, compared with that of C2.

Full jtreg tests on qemu and hotspot/jdk tier1 test on Unmathced are passed without new failures.

JMH tests and results on D1 and Unmatched list as follows:
[MyBenchmark.txt](https://github.com/openjdk/riscv-port/files/7779255/MyBenchmark.txt)

[squareToLen_unmatched.txt](https://github.com/openjdk/riscv-port/files/7779247/squareToLen_unmatched.txt)
[squareToLen_d1.txt](https://github.com/openjdk/riscv-port/files/7779248/squareToLen_d1.txt)
[multiplyToLen_unmatched.txt](https://github.com/openjdk/riscv-port/files/7779249/multiplyToLen_unmatched.txt)
[multiplyToLen_d1.txt](https://github.com/openjdk/riscv-port/files/7779250/multiplyToLen_d1.txt)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279292](https://bugs.openjdk.java.net/browse/JDK-8279292): riscv: Intrinsify multiplyToLen and squareToLen

### Reviewers
 * [Fei Yang](https://openjdk.java.net/census#fyang) (@RealFYang - **Reviewer**)

### Contributors
 * Taiping Guo `<guotaiping1@huawei.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/riscv-port pull/38/head:pull/38` \
`$ git checkout pull/38`

Update a local copy of the PR: \
`$ git checkout pull/38` \
`$ git pull https://git.openjdk.java.net/riscv-port pull/38/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 38`

View PR using the GUI difftool: \
`$ git pr show -t 38`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/riscv-port/pull/38.diff">https://git.openjdk.java.net/riscv-port/pull/38.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 1]
```diff
@@ -3094,6 +3094,479 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
...
+  const Register y_idx = tmp4;
+  const Register carry = tmp5;
+  const Register product = xlen;
+  const Register x_xstart = zlen; // reuse register
+
+  mv(idx, ylen); // idx = ylen;
+  mv(kdx, zlen); // kdx = xlen+ylen;
+  mv(carry, zr); // carry = 0;
+
+  Label L_multiply_64_or_128, L_done;
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Suggestion: rename L_multiply_64_or_128 into L_multiply_64_x_64_loop

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 2]
```diff
@@ -3094,6 +3094,479 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
...
+
+  sub(xstart, xlen, 1);
+  bltz(xstart, L_done);
+
+  const Register jdx = tmp1;
+
+  // if x and y are both 8 bytes aligend.
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: There is a typo here. Maybe "// Check if x and y are both 8-byte aligned."

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 3]
```diff
@@ -3094,6 +3094,479 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
...
+
+  slli(t0, carry2, 32);
+  srli(carry, tmp3, 32);
+  orr(carry, carry, t0);
+
+  bind(L_post_third_loop_done);
+}
+
+/**
+ * Code for BigInteger::multiplyToLen() instrinsic.
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: There is a typo here. Should be: ""Code for BigInteger::multiplyToLen() intrinsic."

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 4]
```diff
@@ -3094,6 +3094,479 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
...
+
+  bind(L_one_x);
+  lwu(x_xstart, Address(x, 0));
+  j(L_first_loop);
+
+  bind(L_first_loop_exit);
+}
+
+/**
+ * Multiply 128 bit by 128. Unrolled inner loop.
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Should be: "Multiply 128 bit by 128 bit. Unrolled inner loop."

**Review by RealFYang — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 1 above)
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Thanks for the suggestion, renamed.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 2 above)
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Oops! Fixed.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 3 above)
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Fixed.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 4 above)
Comment (feilongjiang) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: Fixed.

**Comment by feilongjiang** [comment]
@RealFYang  Thank you for the review. I have addressed all your comments above. :-)

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 5]
```diff
@@ -3094,6 +3094,479 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
...
+  mv(carry, zr); // carry = 0;
+
+  Label L_multiply_64_x_64_loop, L_done;
+
+  sub(xstart, xlen, 1);
+  bltz(xstart, L_done);
+
+  const Register jdx = tmp1;
+
+  // Check if x and y are both 8-byte aligned.
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: I think the code logic for avoiding unaligned access should be under control of option AvoidUnalignedAccesses.

**Review by RealFYang — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp (hunk 5 above)
Comment (guotaiping1) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: The cod logic for avoiding unaligned access has been placed under the control of the AvoidUnalignedAccesses option, and the test cases related to BIgInteger are all passed when `-XX:-AvoidUnalignedAccesses` and `-XX:+AvoidUnalignedAccesses`. Thanks for your review!

**Review by RealFYang — APPROVED** [review]
Looks good.

**Pull request closed without merging**